framework/include/materials/MaterialData.h

```diff
@@ -235,82 +204,57 @@ template<typename T>
 MaterialProperty<T> &
 MaterialData::declareProperty(const std::string & prop_name)
 {
-  unsigned int prop_id = _storage.addProperty(prop_name);
-  resizeProps<T>(prop_id);
-
-  MaterialProperty<T> *prop = dynamic_cast<MaterialProperty<T>*>(_props[prop_id]);
-  mooseAssert(prop != NULL, "Internal error in declaring material property: " + prop_name);
-
-  return *prop;
+  return declareProp<T>(_props, prop_name, _storage.addProperty(prop_name));
 }
 
 template<typename T>
 MaterialProperty<T> &
 MaterialData::declarePropertyOld(const std::string & prop_name)
 {
-  unsigned int prop_id = _storage.addPropertyOld(prop_name);
-  resizeProps<T>(prop_id);
-
-  MaterialProperty<T> *prop = dynamic_cast<MaterialProperty<T>*>(_props_old[prop_id]);
-  mooseAssert(prop != NULL, "Internal error in declaring material property: " + prop_name);
-
-  return *prop;
+  return declareProp<T>(_props_old, prop_name, _storage.addPropertyOld(prop_name));
 }
 
 template<typename T>
 MaterialProperty<T> &
 MaterialData::declarePropertyOlder(const std::string & prop_name)
 {
-  unsigned int prop_id = _storage.addPropertyOlder(prop_name);
-  resizeProps<T>(prop_id);
+  return declareProp<T>(_props_older, prop_name, _storage.addPropertyOlder(prop_name));
+}
 
-  MaterialProperty<T> *prop = dynamic_cast<MaterialProperty<T>*>(_props_older[prop_id]);
+template<typename T>
+MaterialProperty<T> &
+MaterialData::declareProp(MaterialProperties& props, const std::string & prop_name, unsigned int prop_id)
```

Comment:
My two cents, I'm not sure I like seeing both `declareProp()` and `declareProperty()`. I know that the former is private but it's not clear which is which when you are looking through the code. I'd prefer something like `declarePropertyHelper`, or `declarePropertyInternal`